### ISSUE TITLE
MP: Combine Self-Unload mode does not do anything

### ISSUE THREAD

**Issue by jspring0811** [Courseplay/courseplay#4745]
On dedi server with self-unload mode activated, the combine is stopped waiting for an unload (but note there is no cp message indicating "waiting for unload").  I have not tested combine self-unload on a single-player session.

![self-unload1](https://user-images.githubusercontent.com/48362683/71625358-5eeeba80-2bb5-11ea-9bda-744594584939.png)
![self-unload2](https://user-images.githubusercontent.com/48362683/71625364-6746f580-2bb5-11ea-8c95-be17cc2c4f02.png)

**Comment by jspring0811** [comment]
Add...  during this time, all other vehicles on the map are frozen.  I am not able to move any other vehicles until I "stop the driver" on the combine.   When I do click "stop the driver" on the combine, the combine slews forward hundreds of yards (I supose the distance it would have gone while everything was frozen).

**Comment by Tensuko** [comment]
Logfile with debugchannel 7 active.

**Comment by jspring0811** [comment]
Strange...  The dedi "game log" is blank.   My local game log is attached, but it does not have any additional info that I can tell after enabling mode 7 (I enabled mode 7 at about 10:45 in the local log and the combine stopped to unload/froze at about 10:52 in the local log).

Dedi:
![image](https://user-images.githubusercontent.com/48362683/71627407-39b37980-2bc0-11ea-89d4-38b78c4cdc42.png)

Local game log:
[log.txt](https://github.com/Courseplay/courseplay/files/4013057/log.txt)

**Comment by jspring0811** [comment]
... I mean I enabled channel 7
![image](https://user-images.githubusercontent.com/48362683/71627626-8481c100-2bc1-11ea-80e3-737ba141944f.png)

**Comment by Tensuko** [comment]
I am sry but I forgot that we could also need channel 14, that was my bad...

**Comment by pvaiko** [comment]
I can't say much about the MP part of it but that trailer puts the combine in a difficult situation: currently it can only unload on the right side of the trailer which means it needs to get into the fruit to do so. This may result in a very slow pathfinding. You can park the trailers off the field so the combine has enough space on the headland to drive by.

**Comment by pvaiko** [comment]
@davboecki do you know why there's no game log on the server and how to activate debug flags on the server?

**Comment by jspring0811** [comment]
The game freeze occur res no matter where the trailer is parked... here is another instance.  Clicking "Stop driver" on the self-unload Activated combine unfreezes the other vehicles.
![self-unload3](https://user-images.githubusercontent.com/48362683/71628026-ddeaef80-2bc3-11ea-8286-f8051f459bc9.png)
![self-unload4](https://user-images.githubusercontent.com/48362683/71628031-e7745780-2bc3-11ea-9923-d8535be9cf26.png)

**Comment by Tensuko** [comment]
Your Client Log gor a LUA Callstack and then most of the Log is spamed with a Warning, never seen this before.

**Comment by jspring0811** [comment]
I'll setup a new/clean test later today or tomorrow (fresh server restart and test only this scenario with channels 7 and 14 on).

**Comment by jspring0811** [comment]
Ok, I was not able to get the dedi server log fixed yet, so I reproduced the self-unload issue on a local, career-mode game (so non-multiplayer).  I have a warrior semi with open grain trailer sitting at the edge of the field ready to accept the offload.  The results are similar.  When the combine is ready to unload, everything pretty much freezes.  I have attached the game log with debug 7 and 14 enabled.  Issue starts at about 11:45 log time.

Frozen:
![stutter_freeze](https://user-images.githubusercontent.com/48362683/71643931-2d95ed80-2c8e-11ea-9ea5-7876fd5e410e.png)

Log file:
[log.txt](https://github.com/Courseplay/courseplay/files/4014296/log.txt)

**Comment by jspring0811** [comment]
Am I reading this correctly, is the code selecting the combine itself as the best trailer to use for unloading?

2020-01-01 11:45 :03 [dbg14 lp74649] S790 (#1): Fruit left: 0.00 right 0.95, field on left false, right true
2020-01-01 11:45 :03 [dbg14 lp74649] S790 (#1): Semi-Tipper 70000/6 is a trailer on field 0, closest distance to 77 is 2.2, attached to Warrior, root vehicle is Warrior, last speed 0.0
2020-01-01 11:45 :03 [dbg14 lp74649] S790 (#1): S790 is a trailer on field 77, closest distance to 77 is 9.0, attached to none, root vehicle is S790, last speed 8.2
2020-01-01 11:45 :03 [dbg14 lp74649] S790 (#1): HaulMaster is a trailer on field 0, closest distance to 77 is 144.5, attached to AXION 960, root vehicle is AXION 960, last speed 0.0
2020-01-01 11:45 :03 [dbg14 lp74649] S790 (#1): Best trailer is S790 at 0.9 meters, free capacity 1137

**Comment by pvaiko** [comment]
Yes, that's one of the problems, is that a vanilla combine? I have not tested it with all vehicles of course and this one seems to have a Trailer specialization which I can't explain.

**Comment by jspring0811** [comment]
This one is not vanilla, but I was also having trouble with the in-game AGCO combine.   I'll rerun the test with debug 14 and the AGCO IDEAL.

**Comment by power1jj** [comment]
Have the exact same issues using 9230/10.9/S690. Worked great when I first downloaded it but got about 150,000lt’s done and it bugged out.

**Comment by pvaiko** [comment]
356 hopefully fixes the nil error, but why that combine is a trailer is still a mystery.

**Comment by Petje81** [comment]
Not really having the "jump/slew" part as described above but i do have the problem of self unload not working even in v356. On screen the problem looks the same, the combine doesn't fill up completely but thinks it's full... just stops and does nothing. I do get some framedrops when it stops but thats gone after 3 to 5 seconds. Then when i stop CP and start it again from nearest waypoint it does detect the trailer and drives to it.

[20200102_log_v356.txt](https://github.com/Courseplay/courseplay/files/4015754/20200102_log_v356.txt)

PS. the combine i use is modded but does not have a trailer function.

**Comment by pvaiko** [comment]
@Petje81 the 95% and the call stack errors should now be fixed in 357. For the rest I need a log with debug channels 7 and 14.

**Comment by Petje81** [comment]
Got it for you.
I don't know why but i set ch. 7 and 14 before starting the course and it switched to 2 and 19 by itself. I set it to 7 and 14 while working the course.
After the last unload (in the log) there was a hickup, when the combine was empty it shut off and in the console there was nothing happening for a bit. Somehow it found it's way back to working again without me stepping in.

Hope it's of use.

[log.txt](https://github.com/Courseplay/courseplay/files/4015982/log.txt)

**Comment by pvaiko** [comment]
@Petje81 would be more use if you had updated to 357 to get rid of that error and then see if there are any problems left.

**Comment by pvaiko** [comment]
(also, this hybrid A star algorithm I use for pathfinding needs a lot of CPU so depending on the layout, vehicles and fruit around it may take quite a while until it finds the way back)

**Comment by Petje81** [comment]
> @Petje81 would be more use if you had updated to 357 to get rid of that error and then see if there are any problems left.

Sorry, i didn't notice 357 was released already.
While im in the middle of harvest it's no problem to get u another log with 357.

**Comment by jspring0811** [comment]
Fix confirmed: The issues I reported on both Dedi and Local Career mode appear to now be resolved.  Dedi is working now.   On career, the custom modding S790 with incorrect type of "trailer" stops and waits for unload (freezing issue fixed), and other combines self-unload correctly.  Note: self-unload to wilson should be added to the enhancement list, as the combine will drive to the wilson, but then sits there and waits because the front compartment is full and the back compartment is empty.  Same as other wilson enhancement requests.

**Comment by Petje81** [comment]
Log with 357
[log.txt](https://github.com/Courseplay/courseplay/files/4017003/log.txt)

I only finished the field with 357, works pretty good so far, the combine reacts faster for unloading but with 356 it had more trouble (read: took a it more time) finding a trailer again when it got emptied and placed back in a different spot. So ill check on that in a while with a new field.
As for the errors, they look gone from what i could see watching the console.
What i did see was with "return to starting point" enabled the combine stayed at the last unloading point, gave a "reached end point" message and just shut off while still wanting to move. The actual drive speed was 0 while cruise speed was set to 10. You can find this at the end of the log.

**Comment by pvaiko** [comment]
@Petje81 that's intended, not sure a combine would want to go back to the starting point after everything is done and the combine is unloaded. 

I'm closing this here as all the issues seem to be solved.
